### ISSUE TITLE
invariants computed with Pipe algorithm

### ISSUE THREAD

**Issue by yanntm** [CvO-Theory/apt#7]
Hi,
I've been comparing the invariants computed with Pipe to other tools, I find some divergence where Pipe is missing some invariants. More precisely, "test1a" is actually a very bad test, that is forgetting the fact that coefficients used could be negative. 

 I'm looking at models from the model checking contest, such as "Cloud-Deployment-PT-2a".
On this particular model after discarding test1a I can find with ITS-tools : 

invariant :-1'p32 + p33 + p34 + p35 + p36 + p37 + p38 + p39 + p40 + p41 + p42 + p43 + p44 + p45 + p46 + p47 + p48 + p49 + p50 + p51 + p52 + p53 + p54 + p55 + p56 + -1'p57 = 0
invariant :p3 + -1'p57 = 0
invariant :p1 + -1'p57 = 0
invariant :p2 + -1'p32 = 0
invariant :p30 + -1'p32 = 0
invariant :-1'p32 + p67 + p68 = 0
invariant :p4 + -1'p32 = 0
invariant :-1'p32 + p58 + p59 + p60 + p61 + p62 + p65 + p66 = 0
invariant :p5 + p6 + p7 + p8 + p9 + p10 + p11 + p12 + p13 + p14 + p15 + p16 + p17 + p18 + p19 + p20 + p21 + p22 + p23 + p24 + p25 + p26 + p27 + p28 + -1'p32 + -1'p57 = 0
invariant :p29 + -1'p57 = 0
invariant :p63 + p64 + -1'p65 + -1'p66 = 0
invariant :p31 + -1'p57 = 0

Test1a discards and empties the whole matrix pretty fast.

I'm sorry my code base has diverged significantly since I forked from your code, so I can't easily propose a patch.

But I believe this "bug" exists in your code too (ie. your tool does not find the above invariants). It is not a critical bug in the sense that less invariants are reported, but the invariants are all correct. However it's a shame we miss those invariants.

This is the example model in apt format 
[clouDeplPT2a.tar.gz](https://github.com/CvO-Theory/apt/files/3034248/clouDeplPT2a.tar.gz)

**Comment by psychon** [comment]
Hi,

thanks for the report! I'll try to take a look soon-ish, but cannot promise anything.

**Comment by yanntm** [comment]
sure no problem, after comparing to another tool and tracing in my code I concluded this behavior was probably shared with you, so I thought I'd report it here. 
let me know in email or PM if you want help reproducing/comparing with the patched version I have on other examples. I can give you invocation command line examples for my tool so you can compare.
My diverged version lives here, the couple more recent patches relate to fixing this in my code base.
https://github.com/lip6/ITSTools/blob/master/fr.lip6.move.gal.structural/src/uniol/apt/analysis/invariants/InvariantCalculator.java

**Comment by yanntm** [comment]
test 1 a says : if a row (place) has only (in)outgoing edges (only positive or negative) arcs, then forget about it, it cannot belong to a cycle

this is true, in that the place p cannot belong to cycle p->p'->p that is the usual form of invariants

but suppose p' is fed and taken from exactly like p, then p-p'= constant is still true. So subtracting p from the other p' inthe pppm positive or negative set, can still lead to a solution, we should not as test1a suggests give up on p being part of an invariant.

basically this is the root issue that makes the implem miss these invariants in these weird models with redundant places.

**Comment by yanntm** [comment]
so the invariants like p3 + -1'p57 = 0
in my trace are the root things that 1a discards

**Comment by yanntm** [comment]
given that your code always continues to compute only positive invariants, maybe it's actually a non issue for your code base. I break out after general flows are computed, but this whole thing does not make a difference if you go for positive semi flows, since p-p'=cte is not a semiflow

So maybe just close this issue, for the goal of getting semiflows, test1a is good and makes the code faster.

**Comment by prof-best** [comment]
On 4/2/19 at 11:17 PM, notifications@github.com (Yann 
Thierry-Mieg) wrote:

>given that your code always continues to compute only positive invariants, maybe it's actually a non issue for your code base. I break out after general flows are computed, but this whole thing does not make a difference if you go for positive semi flows, since p-p'=cte is not a semiflow
>
>So maybe just close this issue, for the goal of getting semiflows, test1a is good and makes the code faster.

Hello Yann,

I'm not sure I understand correctly, but apt computes _semipositive_
semiflows (invariants) with its "invariants" options.
So, entries could be zero but never negative.
This is written in the brief description which you get by invoking
"apt -jar apt.jar invariants".
If the semipositivity requirement is dropped, there are much faster
algorithms, but I'm not sure any of them is implemented in apt.

Eike

**Comment by yanntm** [comment]
Indeed yes, you are right Eike,

In the tool APT you *output* the semi positive versions, but the algorithm implemented inside is first computing a basis of general flows with negative coefficients possible (this is "phase 1") and it is very fast (polynomial). Phase 2 then tries to rewrite these flows using only positive coefficients, possibly leading to an exponential explosion (so, phase 2 is not guaranteed polynomial). 

So the general algorithm does in fact exist in apt, it's what I based my own implementation off of. Except I break out after phase 1, since the flows are meant to be fed to an SMT solver, that prefers negative coefficients to anything exponential in size. 

The more invariants I get the happier I am however, so losing these p-p'=0 invariants is a shame. But within APT workflow, this (non semi flow) invariant would not be reported anyway.

So I think we should just close this issue, APT tool is fine as it is, and the documentation is accurate. 
We should just keep in mind that a version that would export the flows computed after phase 1 (e.g. a flag for pos or neg invariants) would miss these cases.

**Comment by prof-best** [comment]
On 4/2/19 at 8:20 AM, notifications@github.com (Yann 
Thierry-Mieg) wrote:
>[...] So I think we should just close this issue, ...

Oh, OK, I understand.
Thanks for noticing and documenting this.

**Comment by yanntm** [comment]
The moral I guess is that if you're going for positive semi flows in fine, Test1a is a good test that prunes early in the algorithm invariants that cannot be represented without negative coefficients. If we want general flows, Test1a leads to discarding some of those which might be undesirable.
